public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -789,53 +790,60 @@ export class LokiDatasource
   }
 
   getLogRowContextUi(row: LogRowModel, runContextQuery: () => void): React.ReactNode {
-    return LokiContextUi({
-      row,
-      languageProvider: this.languageProvider,
-      onClose: () => {
-        this.prepareContextExpr = this.prepareContextExprWithoutParsedLabels;
-      },
-      updateFilter: (contextFilters: ContextFilter[]) => {
-        this.prepareContextExpr = async (row: LogRowModel, origQuery?: DataQuery) => {
-          await this.languageProvider.start();
-          const labels = this.languageProvider.getLabelKeys();
-
-          let expr = contextFilters
-            .map((filter) => {
-              const label = filter.value;
-              if (filter && !filter.fromParser && filter.enabled && labels.includes(label)) {
-                // escape backslashes in label as users can't escape them by themselves
-                return `${label}="${escapeLabelValueInExactSelector(row.labels[label])}"`;
-              }
-              return '';
-            })
-            // Filter empty strings
-            .filter((label) => !!label)
-            .join(',');
-
-          expr = `{${expr}}`;
-
-          const parserContextFilters = contextFilters.filter((filter) => filter.fromParser && filter.enabled);
-          if (parserContextFilters.length) {
-            // we should also filter for labels from parsers, let's find the right parser
-            if (origQuery) {
-              const parser = getParserFromQuery((origQuery as LokiQuery).expr);
-              if (parser) {
-                expr = addParserToQuery(expr, parser);
-              }
+    const updateFilter = (contextFilters: ContextFilter[]) => {
```

Comment:
@svennergr could you please share why is this change needed in relation to loading indicator. I am not sure I understand.

Reply:
Sure. Previously the whole component `LokiContextUi` was rerendered on each call. That resulted in different references being generated for the `onClose` property. That resulted in this `onEffect` being called too many times and too early: https://github.com/grafana/grafana/pull/64167/files/2feed00b8817c536231c706617eb0588a0762b55#diff-482a82afb5742195c02896eecc365131b6283b498d80446168cb7daf64286f3cL90
Which then resulted in an too early `clearTimeout` call.

Reply:
gotcha, thanks for context!